### DISCUSSION TITLE
New version of TasmoCompiler - 4.0.0-rc

### DISCUSSION THREAD

**Discussion by benzino77** [benzino77/tasmocompiler#121] [General]
I've just pushed new version of TasmoCompiler (in separate `pa/xversion` branch). There is a lot of changes which are described in commit [notes](https://github.com/benzino77/tasmocompiler/commit/b659c2da9bbaa81614c815d22aaaaadd2c966dad)

Mainly it now supports ESP32 and customization is based on `platformio_override.ini` and `user_config_override.h` file.

I will push docker images soon. For now it can be tested locally or via [gitpod](https://gitpod.io/#https://github.com/benzino77/tasmocompiler/tree/pa/xversion)

**Comment by drhideg** [comment]
Looks good to me. Just a few thoughts so suddenly:
- Afrikaans (`af_AF`), Romanian (`ro_RO`), Vietnamese (`vi_VN`) language option is missing as before.
- `stepFeaturesLightTooltip `has a typo (Lght)
- And my personal whislist with a groupped sensors, where we (I) can select them separate, in a dropdown checkbox in a multiselect, or sorted into groups and show all of them with checkboxes

> &#x25FB; **General features**
> &nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; TASMOTA_CLIENT&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; ADC_VCC&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; USE_PZEM004T **[...]**
> 
> &#x25FB; **Energy sensors**
> &nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; HLW8012&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; CSE7766&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; HOME_ASSISTANT **[...]**
> 
> &#x25FB; **Environment sensors**
> &nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; DS18x20&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; DHT&nbsp;&nbsp;&nbsp;&nbsp;&#x25FB; SHT **[...]**
> 
> **[...]**

**Reply by benzino77** [reply]
First two ware easy to fix and are already there. The third point regarding grouping features... will not be easy to implement, but  I will think about it.

**Comment by benzino77** [comment]
Docker images are pushed to dockerhub. Those should be used as usually are used release candidate versions.

**Comment by drhideg** [comment]
- `stepVersionDesc`: misleading at the moment about core version select
- `stepVersionCore`, `stepVersionBoard`, `stepVersionBoardSpeed`: can be safely removed, not used

PS: You did a good job again. ~~If i figure out how to fork the new branch,~~ i will do the hungarian translation.

**Reply by benzino77** [reply]
There is a lot code cleanup to do. A LOT.
Thanks for finding those places.

**Reply by matrix1000** [reply]
Short Question:
What do you mean, stepVersionBoard and stepVersionBoardSpeed is not used?
I've used it a lot to set Boardversion to Wemos D1 mini
and BoardSpeed to 160 MHZ

Now these options are missing. Why?
I liked the settings. :-)

Very nice and usefull project.  Thanks a lot.

**Reply by drhideg** [reply]
@matrix1000 
Not tested, but I think you can only do that if you modify the source at https://github.com/benzino77/tasmocompiler/blob/21a76e3bf44cd7b68295ceabe1b10967dc026cd1/src/components/AppStepper/FeaturesStep/AvailableBoards.js#L7
to
```javascript
    platformio_entries: {
      'board_build.f_cpu': '160000000L',
    },
```

**Reply by matrix1000** [reply]
Hi, thanks: how can I change it together with gitpod?
And how to change the the flash memory size to 4MB?

**Reply by drhideg** [reply]
Edit `/src/components/AppStepper/FeaturesStep/AvailableBoards.js` as you want then save it. After that, open a new terminal window and run `yarn build` command, then use Open Ports tab and Open Browser again, or just reload the previously opened window to compile tasmota.

ldscrtipt depends on your needs, but if you do not use to store files on spiffs you dont need to modify any
```javascript
    platformio_entries: {
      'board_build.ldscript': 'eagle.flash.4m.ld', //can be eagle.flash.4m.ld, eagle.flash.4m1m.ld, eagle.flash.4m2m.ld, eagle.flash.4m3m.ld
      'board_build.f_cpu': '160000000L',
    },
```

**Reply by benzino77** [reply]
https://gitpod.io/#https://github.com/benzino77/tasmocompiler/commit/3ad48c86d128f6ad18e3274ff5307ce5fbc4b23c

This is the link to run TasmoCompiler 3.1.8 on Gitpod.

**Reply by matrix1000** [reply]
Many Thanks :-)

**Comment by rapi3** [comment]
Hi,
Is there now any possibility to compile for Wemos D1 mini for UFS support with:
speed 160Mhz
board_build.ldscript = eagle.flash.4m2m.ld
thank you

**Comment by benzino77** [comment]
I have to rethink removal of board type :?). You don't need "boosting" board speed for UFS (I think).
For now, you can use last 3.x version of TC which is available on dockerhub or can be launched on gitpod via this link:
https://gitpod.io/#https://github.com/benzino77/tasmocompiler/commit/3ad48c86d128f6ad18e3274ff5307ce5fbc4b23c

**Reply by rapi3** [reply]
I try to compile my fork from latest development for wemos D1 mini 160Mhz with UFS using v3 from link and it fail to have that option after succesfully compiled.
Any idea ??

here it is the platform.io downloaded

`[platformio]
src_dir=tasmota
lib_dir=lib/default
build_cache_dir=.cache
default_envs=firmware
extra_configs=tc_cores.ini

[user_defined]
board_memory=eagle.flash.4m1m.ld
board=d1_mini
f_cpu=160000000L
build_flags=

[core_active]
platform=${core_2_7_4.platform}
platform_packages=${core_2_7_4.platform_packages}
build_flags=${core_2_7_4.build_flags}

[common]
framework=arduino
board=${user_defined.board}
board_build.flash_mode=dout
platform=${core_active.platform}
platform_packages=${core_active.platform_packages}
build_flags=${core_active.build_flags} ${user_defined.build_flags} -DUSE_CONFIG_OVERRIDE
build_unflags=-Wall
board_build.f_cpu=${user_defined.f_cpu}
board_build.f_flash=40000000L
upload_resetmethod=nodemcu
extra_scripts=pio-tools/strip-floats.py, pio-tools/name-firmware.py, pio-tools/gzip-firmware.py, pio-tools/override_copy.py
lib_ldf_mode=chain+
shared_libdeps_dir=lib
lib_extra_dirs=lib/lib_basic, lib/lib_i2c, lib/lib_display, lib/lib_ssl, lib/lib_audio, lib/lib_rf, lib/lib_div

[env:firmware]
platform=${common.platform}
platform_packages=${common.platform_packages}
framework=${common.framework}
board=${common.board}
board_build.ldscript=${user_defined.board_memory}
board_build.flash_mode=${common.board_build.flash_mode}
board_build.f_cpu=${common.board_build.f_cpu}
build_unflags=${common.build_unflags}
build_flags=${common.build_flags}
upload_resetmethod=${common.upload_resetmethod}
extra_scripts=${common.extra_scripts}
lib_ldf_mode=${common.lib_ldf_mode}
lib_extra_dirs=${common.lib_extra_dirs}`

**Comment by benzino77** [comment]
Maybe it is not able to find linker script you provided. Try with this one:
`board_build.ldscript        = eagle.flash.4m2m.ld`

You have to remember that Tasmota is evolving very fast. Especially in platformio/compiler/available boards area. This is not very consistent (this is my opinion) and change from release to release. `platformio*.ini` files are changing over and over again. From side project perspective (like TasmoCompiler) it's a **nightmare**. This is not complaining. Tasmota dev team is doing its job, moving the project forward in their spare time. I'm just saying that it's hard to keep some functionality in TasmoCompiler because of changes in Tasmota itself.

Anyway I will try to give the opportunity to choose 4Mb boards and add UFS as a option to choose.

**Reply by rapi3** [reply]
I already have that line activated in platformio_override.ini and platformio_override_sample.ini in my custom fork.
Do I need to activate script for this to work or anything else because I am using custom options and disabled all sensor and stuff I don't need to have a low ram usage ?

**Comment by benzino77** [comment]
`#define USE_UFILESYS`
and
![obraz](https://user-images.githubusercontent.com/15658235/113750674-6a5f5880-970b-11eb-9285-7353efbf3cf3.png)


Compiles well on my side.

**Reply by rapi3** [reply]
ok I will try again, I was missing that define line

**Reply by rapi3** [reply]
All good now using v3 and my custom fork.
thank you.

**Comment by benzino77** [comment]
This week I will try to release v4.x version which supports 4Mb boards (like Wemos or NodeMCU) and add UFS as a feature to select.

**Reply by rapi3** [reply]
Maybe you can add something to have the /Tasmota folder visible from gitpod so we can make small changes on downloaded files before starting to compile.

**Reply by benzino77** [reply]
That's gitpod functionality. You have to search on gitpod help/docs.

**Comment by benzino77** [comment]
It's looks promising (compiled with Tasmota 4.x and uploaded to Wemos D1 mini):
![ufs01](https://user-images.githubusercontent.com/15658235/113759420-56205900-9715-11eb-9807-2b08b0105dc1.png)
![obraz](https://user-images.githubusercontent.com/15658235/113759606-89fb7e80-9715-11eb-9d03-de8f8c4b3a55.png)

**Reply by rapi3** [reply]
Something it is not ok.
I compiled my fork with Tasmocompiler and look like it ignore my settings on platformio_override.ini because the result show only 1Mb UFS flash.
![Screenshot_2021-04-06_22-28-12](https://user-images.githubusercontent.com/18531150/113774539-b5876480-9727-11eb-964f-cc93145fd3b4.png)

compiled my fork directly in gitpod will compile correctly with 2Mb UFS flash
![Screenshot_2021-04-06_19-14-54](https://user-images.githubusercontent.com/18531150/113774653-d64fba00-9727-11eb-8450-70d6f1538c92.png)

**Reply by rapi3** [reply]
solved, switched to 4rc -forked, patched AvailableBoards.js and ok

**Comment by benzino77** [comment]
New version 4.1.0 is published with  4Mb boards and UFILESYS support. You can use it on gitpod regular way. Choose `Wemos/NodeMCU` as your board and `SD card/LittleFS` on features list and you should be good to go.

**Reply by rapi3** [reply]
How it is the Frequency set as is no more user configurable ? 80 or 160MHz ?

**Comment by benzino77** [comment]
There is no need to set clock speed for that. If there is a need to change clock speed for example for `Sonoff Zigbee Bridge` or for `ESP32 WebCam` TasmoCompiler is doing it under the hood. 
You don't need to change clock speed for `Universal File System` (SD card or memory based).

If you need to change it anyway you have to use older version of TC or prepare your own fork.

**Reply by rapi3** [reply]
Yes and the frequency for wemos it is set at ?

**Reply by benzino77** [reply]
@drhideg gives you an example few posts earlier:  https://github.com/benzino77/tasmocompiler/discussions/121#discussioncomment-569851

**Reply by rapi3** [reply]
The short answer it is Frequency is set by default at 80Mz.

**Reply by benzino77** [reply]
Yes, that's default frequency Tasmota is using.

**Reply by matrix1000** [reply]
Is the higher cpu speed not usefull for faster webinterface, other calculations and reactions?

**Reply by rapi3** [reply]
I am using all Wemos D1, NODE MCU, Tuya plugs, Sonoff plugs, rf bridge...I have in the house running Tasmota at 160Mhz without any problem.
So for me it is important to stay on 160Mhz

**Reply by benzino77** [reply]
I don't think so. If it would, Tasmota Dev Team would make higher frequency a default value for all their prebuild binaries. It is boosted only in situations where it is needed.

**Reply by rapi3** [reply]
running at 160Mhz you can see clearly in console a shorten response time at connecting if you use MQTT TLS.
Web interface it is more responsive also.
Everything run smoother at higher freq on my devices.

**Reply by benzino77** [reply]
> I am using all Wemos D1, NODE MCU, Tuya plugs, Sonoff plugs, rf bridge...I have in the house running Tasmota at 160Mhz without any problem.
> So for me it is important to stay on 160Mhz

Devices compiled with clock speed 160MHz and 80Mhz are **totally compatible**. There is no need to boost frequency for every device (especially that device working with higher frequency will consume more energy). You can have devices working with 80Mhz/160Mhz or for ESP 32 160Mhz/240Mhz in your ecosystem.

**Reply by benzino77** [reply]
> running at 160Mhz you can see clearly in console a shorten response time at connecting if you use MQTT TLS.
> Web interface it is more responsive also.
> Everything run smoother at higher freq on my devices.

https://tasmota.github.io/docs/TLS/#connection-time

It's fraction of second. The biggest different is 0.8s. You are connecting to broker only on startup or when you lost WiFi connection. In day to day usage you will not log to Web Console. You will use NodeRed/HomeAssistant/Domoticz/Rules/Scripts etc. to do the job.

To summarize that discussion. You have couple of options:
1. You can use older version of TasmoCompiler (but there is a chance that it will not working with newer version of Tasmota)
2. Fork TasmoCompiler repo and create your own version (with `'board_build.f_cpu': '160000000L'`)
3. Contact Tasmota Dev Team and convince them that 160Mhz should be default for ESP8266
4. Use default settings as it will not impact your overall performance/functionality/etc.

**Reply by Jason2866** [reply]
ESP8266 standard MCU freq. will not be changed, there is no need for. Tasmota runs well.
Not all devices out there will run stable at 160Mhz. The get hotter and draw more power.
For using TLS you have to compile yourself anyway so changing to 160Mhz if you want (no need) is just a simple config change.
@benzino77  i know changing platformio setup is a nightmare for side projects of Tasmota.
BUT only changed when new features and functions are added. Platformio is a part of Tasmota.
So the setup can not be consistent. The filesystem UFS is a example for. Nobody is forced to use the latest version of Tasmota :-)
Without PlatformIO Tasmota would not be possible as it is now!!

**Reply by benzino77** [reply]
Sure, it's completely understandable for me. I'm not complaining at all... maybe a little :) 
I'm just saying that Tasmota is moving forward very fast, and keeping consistency with such fast moving projects is hard to follow for projects with one or two developers like TC.

Totally agree that there is no need to "boost" frequency for situations where it is not needed.
